lab3/middlebox.py

```diff
@@ -0,0 +1,139 @@
+#!/usr/bin/env python3
+
+"""
+David to implement
+"""
+
+from switchyard.lib.address import *
+from switchyard.lib.packet import *
+from switchyard.lib.packet.util import *
+from switchyard.lib.userlib import *
+from threading import *
+import random
+import time
+
+def drop(percent):
+    return random.randrange(100) < percent
+
+def delay(mean, std):
+    delay =random.gauss(mean, std)
+    print(delay)
+    if delay > 0:
+        time.sleep(delay/1000)
+
+'''
+load_file
+Loads the configuration file for the middlebox.
+Parameters:
+    filename    - name of the file to load
+Returns:
+    dictionary with the following fields:
+        s     = pseudorandom seed
+        p     = probability of packet drop
+        dm    = mean delay time
+        dstd  = mean standard deviation
+'''
+def load_file(filename):
+    log_debug("Loading file: {}".format(filename))
+
+    config = {}
+    try:
+        fh = open(filename, 'r')
+        line = fh.read().replace('\n', '').strip()
+        fh.close()
+    except:
+        log_debug("Failed to load config file {}: {}".format(filename, sys.exc_info()))
+        return config
+
+    fields = line.split()
+    if (len(fields) != 8):
+        log_debug("Bad file contents: {}".format(line))
+        return config
+    for i in range(0, 7, 2):
+        key = fields[i].replace('-', '')
+        val = fields[i+1]
+        config[key] = int(val)
+
+    return config
+
+'''
+update_pkt
+Replaces the Ethernet header when forwarding.
+Updates ttl on the IPv4 header
+Parameters:
+    pkt         - packet to update
+    out_port    - output port (EthAddr)
+    out_dst     - dest mac (EthAddr)
+'''
+def update_pkt(pkt, out_port, out_dst):
+    del pkt[0]
+    eth_head = Ethernet(src=out_port, dst=out_dst, ethertype=EtherType.IPv4)
+    pkt.prepend_header(eth_head)
+
+    #Decrement TTL
+    pkt[IPv4].ttl -= 1
+
+
+def switchy_main(net):
+
+    my_intf = net.interfaces()
+    mymacs = [intf.ethaddr for intf in my_intf]
+    myips = [intf.ipaddr for intf in my_intf]
+
+    config = load_file("./middlebox_params.txt")
+
+    random_seed = config['s']
+    random.seed(random_seed) #Extract random seed from params file
+
+    while True:
+        gotpkt = True
```

Comment:
We can obviate the need for this `gotpkt` state tracking variable if you replace the `except NoPackets:` block with this:

```

        except NoPackets:
            log_debug("No packets available in recv_packet")
            continue 
```

With the current code, even in the no-packet case, line through 103 through 137 will run. These lines shouldn't be run in the no-packet case.